qiskit/quantum_info/synthesis/xx_decompose/polytopes.py

```diff
@@ -0,0 +1,266 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Defines bare dataclasses which house polytope information, as well as a specialized data structure
+which describes those two-qubit programs accessible to a given sequence of XX-type interactions.
+"""
+
+from copy import copy
+from dataclasses import dataclass, field
```

Comment:
The CI is not testing Python 3.6 and support for 3.6 was marked as deprecated some time ago. 
If we really need this release to be compatible then i would just do a `try` before the import here and raise a meaningful error. It's not a core feature of Qiskit so I'm not worried about it not being accessible to a subset of users.

Reply:
It's not running on 3.6 here yet because that's part of stage 2 in the CI pipeline which tests python 3.6, 3.8, and 3.9 on all operating systems (along with testing the tutorials). The first stage only runs with 3.7 instead of everything to improve throughput if a PR is broken and eats up ci telling us that 10 more times. This PR is failing in the first stage so the second stage which tests all the other python versions hasn't run yet. I expect that @levbishop is correct and once this gets past the first stage of CI it will fail on python 3.6 because of the dataclasses usage.

0.19.x is the last release where we support Python 3.6 (which has been advertised a bunch of places already and for some time) so we need to make sure things still work with it. After 0.19.0 is released I plan to push a PR raising the minimum python version to 3.7 and drop the 3.6 CI jobs. If you need to use dataclasses here for the 0.19.0 release then for python 3.6 you can do this with the backport package: https://pypi.org/project/dataclasses/ or as you suggest raise a informative error message on Python 3.6 when it fails trying to import dataclasses, however this is contingent on this module only being imported when the feature is used (if the module gets imported by other things this may not be the case).

Reply:
Rather than pull out dataclasses only to add it back immediately after the release, i prefer @mtreinish idea to use the backport. I believe that should be as easy as adding a line to requirements.txt
```
dataclasses>=0.8;python_version<'3.7'
```
and making a mental note to remove it once 0.19 ships

Reply:
Oh: I was under the impression that this required copying the dependency into the source tree (since I once saw someone else do that). If that suffices, then that's my preference too!